File: hooks/python/periskop_hook/blake3.py
"""BLAKE3, written out rather than installed.

data-model.md section 2 fixes one hash for every identity in this project and
names it: blake3, first 8 bytes, 16 lowercase hex. It rejects sha256 by name, so
`hashlib` does not answer the question: CPython ships sha2, blake2b and sha3, and
none of them is blake3. PyPI does, but this package is imported into somebody
else's production interpreter through a `.pth` file, and every distribution it
drags in is one that process is now forced to install and trust. The identity
contract and the zero dependency rule meet here, and this file is the meeting
point.

This is a line for line port of `hooks/node/src/blake3.ts` so that a reader can
diff the two and see that they are the same algorithm. Both are checked against
the same reference vectors (`tests/test_blake3.py`), which is what makes a python
hook and a node hook agree on the identity of one call.

Scope: the 32 byte, unkeyed hash. Keyed hashing and key derivation are not
implemented because no identity in this project uses them.
"""

import struct

_OUT_LEN = 32
_BLOCK_LEN = 64
_CHUNK_LEN = 1024

_CHUNK_START = 1
_CHUNK_END = 2
_PARENT = 4
_ROOT = 8

# Python integers do not wrap, so every arithmetic step masks back to 32 bits.
_MASK32 = 0xFFFFFFFF

_IV = (
    0x6A09E667, 0xBB67AE85, 0x3C6EF372, 0xA54FF53A,
    0x510E527F, 0x9B05688C, 0x1F83D9AB, 0x5BE0CD19,
)

_MSG_PERMUTATION = (2, 6, 3, 10, 7, 0, 4, 13, 1, 11, 12, 5, 9, 14, 15, 8)

_WORDS_FORMAT = struct.Struct("<16I")


def _rotr(value, bits):
    return ((value >> bits) | (value << (32 - bits))) & _MASK32


def _mix(state, a, b, c, d, mx, my):
    state[a] = (state[a] + state[b] + mx) & _MASK32
    state[d] = _rotr(state[d] ^ state[a], 16)
    state[c] = (state[c] + state[d]) & _MASK32
    state[b] = _rotr(state[b] ^ state[c], 12)
    state[a] = (state[a] + state[b] + my) & _MASK32
    state[d] = _rotr(state[d] ^ state[a], 8)
    state[c] = (state[c] + state[d]) & _MASK32
    state[b] = _rotr(state[b] ^ state[c], 7)


def _round(state, m):
    _mix(state, 0, 4, 8, 12, m[0], m[1])
    _mix(state, 1, 5, 9, 13, m[2], m[3])
    _mix(state, 2, 6, 10, 14, m[4], m[5])
    _mix(state, 3, 7, 11, 15, m[6], m[7])
    _mix(state, 0, 5, 10, 15, m[8], m[9])
    _mix(state, 1, 6, 11, 12, m[10], m[11])
    _mix(state, 2, 7, 8, 13, m[12], m[13])
    _mix(state, 3, 4, 9, 14, m[14], m[15])


def _compress(chaining, block_words, counter, block_len, flags):
    state = [
        chaining[0], chaining[1], chaining[2], chaining[3],
        chaining[4], chaining[5], chaining[6], chaining[7],
        _IV[0], _IV[1], _IV[2], _IV[3],
        counter & _MASK32,
        (counter >> 32) & _MASK32,
        block_len,
        flags,
    ]
    block = list(block_words)
    for index in range(7):
        _round(state, block)
        if index < 6:
            block = [block[position] for position in _MSG_PERMUTATION]

    for i in range(8):
        state[i] ^= state[i + 8]
        state[i + 8] ^= chaining[i]
    return state


def _block_to_words(block):
    """The 64 byte block as 16 little endian words."""
    return list(_WORDS_FORMAT.unpack(bytes(block)))


class _Output(object):
    """A node of the tree, not yet told whether it is the root."""

    __slots__ = ("chaining", "block_words", "counter", "block_len", "flags")

    def __init__(self, chaining, block_words, counter, block_len, flags):
        self.chaining = chaining
        self.block_words = block_words
        self.counter = counter
        self.block_len = block_len
        self.flags = flags


def _chaining_value_of(output):
    return _compress(
        output.chaining, output.block_words, output.counter,
        output.block_len, output.flags,
    )[:8]


def _root_bytes(output):
    # Only the first 64 output bytes are reachable at counter zero, and 32 is
    # all the identity format asks for, so the extended output stream is not
    # built.
    words = _compress(
        output.chaining, output.block_words, 0,
        output.block_len, output.flags | _ROOT,
    )
    return struct.pack("<8I", *words[:8])[:_OUT_LEN]


def _parent_output(left, right, flags):
    return _Output(
        chaining=_IV,
        block_words=list(left) + list(right),
        counter=0,
        block_len=_BLOCK_LEN,
        flags=flags | _PARENT,
    )


class _ChunkState(object):
    __slots__ = ("_chaining", "_block", "_block_len", "_blocks_compressed",
                 "counter", "_flags")

    def __init__(self, chaining, counter, flags):
        self._chaining = list(chaining[:8])
        self._block = bytearray(_BLOCK_LEN)
        self._block_len = 0
        self._blocks_compressed = 0
        self.counter = counter
        self._flags = flags

    def length(self):
        return _BLOCK_LEN * self._blocks_compressed + self._block_len

    def _start_flag(self):
        return _CHUNK_START if self._blocks_compressed == 0 else 0

    def update(self, data):
        offset = 0
        size = len(data)
        while offset < size:
            if self._block_len == _BLOCK_LEN:
                self._chaining = _compress(
                    self._chaining,
                    _block_to_words(self._block),
                    self.counter,
                    _BLOCK_LEN,
                    self._flags | self._start_flag(),
                )[:8]
                self._blocks_compressed += 1
                self._block = bytearray(_BLOCK_LEN)
                self._block_len = 0
            take = min(_BLOCK_LEN - self._block_len, size - offset)
            self._block[self._block_len:self._block_len + take] = (
                data[offset:offset + take])
            self._block_len += take
            offset += take

    def output(self):
        return _Output(
            chaining=self._chaining,
            block_words=_block_to_words(self._block),
            counter=self.counter,
            block_len=self._block_len,
            flags=self._flags | self._start_flag() | _CHUNK_END,
        )


class _Hasher(object):
    __slots__ = ("_chunk", "_stack")

    def __init__(self):
        self._chunk = _ChunkState(_IV, 0, 0)
        self._stack = []

    def update(self, data):
        offset = 0
        size = len(data)
        while offset < size:
            if self._chunk.length() == _CHUNK_LEN:
                chunk_cv = _chaining_value_of(self._chunk.output())
                total_chunks = self._chunk.counter + 1
                self._merge_chunk(chunk_cv, total_chunks)
                self._chunk = _ChunkState(_IV, total_chunks, 0)
            take = min(_CHUNK_LEN - self._chunk.length(), size - offset)
            self._chunk.update(data[offset:offset + take])
            offset += take

    def _merge_chunk(self, chunk_cv, total_chunks):
        """Merge a completed chunk with everything on the stack it completes.

        The number of trailing zero bits of the chunk count says how many merges
        are due, which is what keeps the stack logarithmic in the input length.
        """
        chaining = chunk_cv
        chunks = total_chunks
        while chunks % 2 == 0:
            if not self._stack:
                break
            left = self._stack.pop()
            chaining = _chaining_value_of(_parent_output(left, chaining, 0))
            chunks //= 2
        self._stack.append(chaining)

    def digest(self):
        output = self._chunk.output()
        for index in range(len(self._stack) - 1, -1, -1):
            output = _parent_output(
                self._stack[index], _chaining_value_of(output), 0)
        return _root_bytes(output)


def blake3(data):
    """32 byte unkeyed BLAKE3 digest of the given bytes."""
    hasher = _Hasher()
    hasher.update(data)
    return hasher.digest()


def blake3_short(data):
    """The short form every periskop identity uses: first 8 bytes, lower hex."""
    return blake3(data)[:8].hex()

File: hooks/node/src/blake3.ts
// BLAKE3, written out rather than installed.
//
// data-model.md section 2 fixes one hash for every identity in this project and
// names it: blake3, first 8 bytes, 16 lowercase hex. sha256 is rejected there by
// name, so node:crypto does not answer the question. npm does, but a hook is
// loaded into somebody else's production process, and every package it drags in
// is a package that process is now forced to trust. The identity contract and
// the zero-dependency rule meet here, and this file is the meeting point.
//
// Scope: the 32 byte, unkeyed hash. Keyed hashing and key derivation are not
// implemented because no identity in this project uses them.

const OUT_LEN = 32;
const BLOCK_LEN = 64;
const CHUNK_LEN = 1024;

const CHUNK_START = 1;
const CHUNK_END = 2;
const PARENT = 4;
const ROOT = 8;

const IV = new Uint32Array([
  0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19,
]);

const MSG_PERMUTATION = new Uint8Array([2, 6, 3, 10, 7, 0, 4, 13, 1, 11, 12, 5, 9, 14, 15, 8]);

// Every index below is in bounds by construction: the arrays are fixed length
// and the loops are fixed count. The accessor keeps noUncheckedIndexedAccess
// from spraying assertions through the round function, where they would hide
// the shape of the algorithm.
function word(words: Uint32Array, index: number): number {
  return words[index] as number;
}

function byte(bytes: Uint8Array, index: number): number {
  return bytes[index] as number;
}

function rotr(value: number, bits: number): number {
  return ((value >>> bits) | (value << (32 - bits))) >>> 0;
}

function mix(
  state: Uint32Array,
  a: number,
  b: number,
  c: number,
  d: number,
  mx: number,
  my: number,
): void {
  state[a] = (word(state, a) + word(state, b) + mx) >>> 0;
  state[d] = rotr(word(state, d) ^ word(state, a), 16);
  state[c] = (word(state, c) + word(state, d)) >>> 0;
  state[b] = rotr(word(state, b) ^ word(state, c), 12);
  state[a] = (word(state, a) + word(state, b) + my) >>> 0;
  state[d] = rotr(word(state, d) ^ word(state, a), 8);
  state[c] = (word(state, c) + word(state, d)) >>> 0;
  state[b] = rotr(word(state, b) ^ word(state, c), 7);
}

function round(state: Uint32Array, m: Uint32Array): void {
  mix(state, 0, 4, 8, 12, word(m, 0), word(m, 1));
  mix(state, 1, 5, 9, 13, word(m, 2), word(m, 3));
  mix(state, 2, 6, 10, 14, word(m, 4), word(m, 5));
  mix(state, 3, 7, 11, 15, word(m, 6), word(m, 7));
  mix(state, 0, 5, 10, 15, word(m, 8), word(m, 9));
  mix(state, 1, 6, 11, 12, word(m, 10), word(m, 11));
  mix(state, 2, 7, 8, 13, word(m, 12), word(m, 13));
  mix(state, 3, 4, 9, 14, word(m, 14), word(m, 15));
}

function permute(m: Uint32Array): void {
  const permuted = new Uint32Array(16);
  for (let i = 0; i < 16; i += 1) permuted[i] = word(m, byte(MSG_PERMUTATION, i));
  m.set(permuted);
}

function compress(
  chainingValue: Uint32Array,
  blockWords: Uint32Array,
  counter: number,
  blockLen: number,
  flags: number,
): Uint32Array {
  const state = new Uint32Array(16);
  state.set(chainingValue.subarray(0, 8), 0);
  state.set(IV.subarray(0, 4), 8);
  state[12] = counter >>> 0;
  state[13] = Math.floor(counter / 0x1_0000_0000) >>> 0;
  state[14] = blockLen;
  state[15] = flags;

  const block = Uint32Array.from(blockWords);
  for (let r = 0; r < 7; r += 1) {
    round(state, block);
    if (r < 6) permute(block);
  }

  for (let i = 0; i < 8; i += 1) {
    state[i] = (word(state, i) ^ word(state, i + 8)) >>> 0;
    state[i + 8] = (word(state, i + 8) ^ word(chainingValue, i)) >>> 0;
  }
  return state;
}

function blockToWords(block: Uint8Array, words: Uint32Array): void {
  for (let i = 0; i < 16; i += 1) {
    const at = i * 4;
    words[i] =
      (byte(block, at) |
        (byte(block, at + 1) << 8) |
        (byte(block, at + 2) << 16) |
        (byte(block, at + 3) << 24)) >>>
      0;
  }
}

/** A node of the tree, not yet told whether it is the root. */
interface Output {
  readonly chaining: Uint32Array;
  readonly blockWords: Uint32Array;
  readonly counter: number;
  readonly blockLen: number;
  readonly flags: number;
}

function chainingValueOf(output: Output): Uint32Array {
  return compress(
    output.chaining,
    output.blockWords,
    output.counter,
    output.blockLen,
    output.flags,
  ).slice(0, 8);
}

function rootBytes(output: Output): Uint8Array {
  // Only the first 64 output bytes are reachable at counter zero, and 32 is all
  // the identity format asks for, so the extended output stream is not built.
  const words = compress(output.chaining, output.blockWords, 0, output.blockLen, output.flags | ROOT);
  const bytes = new Uint8Array(OUT_LEN);
  for (let i = 0; i < 8; i += 1) {
    const value = word(words, i);
    bytes[i * 4] = value & 0xff;
    bytes[i * 4 + 1] = (value >>> 8) & 0xff;
    bytes[i * 4 + 2] = (value >>> 16) & 0xff;
    bytes[i * 4 + 3] = (value >>> 24) & 0xff;
  }
  return bytes;
}

function parentOutput(left: Uint32Array, right: Uint32Array, flags: number): Output {
  const blockWords = new Uint32Array(16);
  blockWords.set(left, 0);
  blockWords.set(right, 8);
  return { chaining: IV, blockWords, counter: 0, blockLen: BLOCK_LEN, flags: flags | PARENT };
}

class ChunkState {
  #chaining: Uint32Array;
  readonly #block = new Uint8Array(BLOCK_LEN);
  readonly #words = new Uint32Array(16);
  #blockLen = 0;
  #blocksCompressed = 0;

  constructor(
    chaining: Uint32Array,
    readonly counter: number,
    private readonly flags: number,
  ) {
    this.#chaining = chaining.slice(0, 8);
  }

  length(): number {
    return BLOCK_LEN * this.#blocksCompressed + this.#blockLen;
  }

  #startFlag(): number {
    return this.#blocksCompressed === 0 ? CHUNK_START : 0;
  }

  update(input: Uint8Array): void {
    let offset = 0;
    while (offset < input.length) {
      if (this.#blockLen === BLOCK_LEN) {
        blockToWords(this.#block, this.#words);
        this.#chaining = compress(
          this.#chaining,
          this.#words,
          this.counter,
          BLOCK_LEN,
          this.flags | this.#startFlag(),
        ).slice(0, 8);
        this.#blocksCompressed += 1;
        this.#block.fill(0);
        this.#blockLen = 0;
      }
      const take = Math.min(BLOCK_LEN - this.#blockLen, input.length - offset);
      this.#block.set(input.subarray(offset, offset + take), this.#blockLen);
      this.#blockLen += take;
      offset += take;
    }
  }

  output(): Output {
    const blockWords = new Uint32Array(16);
    blockToWords(this.#block, blockWords);
    return {
      chaining: this.#chaining,
      blockWords,
      counter: this.counter,
      blockLen: this.#blockLen,
      flags: this.flags | this.#startFlag() | CHUNK_END,
    };
  }
}

class Hasher {
  #chunk = new ChunkState(IV, 0, 0);
  readonly #stack: Uint32Array[] = [];

  update(input: Uint8Array): void {
    let offset = 0;
    while (offset < input.length) {
      if (this.#chunk.length() === CHUNK_LEN) {
        const chunkCv = chainingValueOf(this.#chunk.output());
        const totalChunks = this.#chunk.counter + 1;
        this.#mergeChunk(chunkCv, totalChunks);
        this.#chunk = new ChunkState(IV, totalChunks, 0);
      }
      const take = Math.min(CHUNK_LEN - this.#chunk.length(), input.length - offset);
      this.#chunk.update(input.subarray(offset, offset + take));
      offset += take;
    }
  }

  // A completed chunk merges with everything on the stack that it completes.
  // The number of trailing zero bits of the chunk count says how many merges
  // are due, which is what keeps the stack logarithmic in the input length.
  #mergeChunk(chunkCv: Uint32Array, totalChunks: number): void {
    let cv = chunkCv;
    let chunks = totalChunks;
    while (chunks % 2 === 0) {
      const left = this.#stack.pop();
      if (left === undefined) break;
      cv = chainingValueOf(parentOutput(left, cv, 0));
      chunks = Math.floor(chunks / 2);
    }
    this.#stack.push(cv);
  }

  digest(): Uint8Array {
    let output = this.#chunk.output();
    for (let i = this.#stack.length - 1; i >= 0; i -= 1) {
      output = parentOutput(this.#stack[i] as Uint32Array, chainingValueOf(output), 0);
    }
    return rootBytes(output);
  }
}

/** 32 byte unkeyed BLAKE3 digest of the given bytes. */
export function blake3(input: Uint8Array): Uint8Array {
  const hasher = new Hasher();
  hasher.update(input);
  return hasher.digest();
}

/** The short form every periskop identity uses: first 8 bytes, lowercase hex. */
export function blake3Short(input: Uint8Array): string {
  const digest = blake3(input);
  let hex = "";
  for (let i = 0; i < 8; i += 1) hex += byte(digest, i).toString(16).padStart(2, "0");
  return hex;
}
